openssl/src/x509/mod.rs

```diff
@@ -1322,6 +1322,32 @@ impl X509AlgorithmRef {
     }
 }
 
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::X509_OBJECT;
+    fn drop = ffi::X509_OBJECT_free;
+
+    /// An `X509` or an X509 certificate revocation list.
+    pub struct X509Object;
+    /// Reference to `X509Object`
+    pub struct X509ObjectRef;
+}
+
+impl X509ObjectRef {
+    pub fn x509(&self) -> Option<&X509Ref> {
+        unsafe {
+            if X509_OBJECT_get_type(self.as_ptr()) == ffi::X509_LU_X509 {
+                Some(X509Ref::from_ptr(X509_OBJECT_get0_X509(self.as_ptr())))
```

Comment:
X509_OBJECT_get0_X509 returns null if the object isn't a certificate, so you should be able to skip the X509_OBJECT_get_type call and just check the pointer.